substrate-node/pallets/pallet-smart-contract/src/migrations/v9.rs

```diff
@@ -1,194 +1,958 @@
+        9 => (clean_contract_billing_information_by_id::<T>(), Some(current_stage + 1)),
+        // Last cleaning operation, set stage to none to stop migration
+        10 => {
+            let weight = clean_node_contract_resources::<T>();
+            PalletVersion::<T>::put(types::StorageVersion::V9);
+            info!("🔔 Ending Smart Contract pallet storage cleaning");
+            (weight.add(1), None)
+        }
+        // Should never happen
+        _ => (Weight::zero(), None),
+    }
+}
+
+// Contracts
+pub fn check_contracts<T: Config>() {
+    debug!(
+        "🔎  Smart Contract pallet {:?} checking Contracts storage map START",
+        PalletVersion::<T>::get()
+    );
+
+    let contract_id_range = 1..=ContractID::<T>::get();
+
+    for (contract_id, contract) in Contracts::<T>::iter() {
+        if contract_id != contract.contract_id {
+            debug!(
+                " ⚠️    Contracts[id: {}]: wrong id ({})",
+                contract_id, contract.contract_id
+            );
+        }
+        if !contract_id_range.contains(&contract_id) {
+            debug!(
+                " ⚠️    Contracts[id: {}]: id not in range {:?}",
+                contract_id, contract_id_range
+            );
+        }
+
+        match contract.contract_type {
+            types::ContractData::NodeContract(node_contract) => check_node_contract::<T>(
+                node_contract.node_id,
+                contract_id,
+                node_contract.deployment_hash,
+            ),
+            types::ContractData::NameContract(name_contract) => {
+                check_name_contract::<T>(contract_id, &name_contract.name)
+            }
+            types::ContractData::RentContract(ref rent_contract) => {
+                check_rent_contract::<T>(rent_contract.node_id, &contract)
+            }
+        }
+
+        // ContractLock
+        if !ContractLock::<T>::contains_key(contract_id) {
+            debug!(
+                " ⚠️    Contract (id: {}): no contract lock found",
+                contract_id
+            );
+        }
+
+        // SolutionProviders
+        if let Some(sp_id) = contract.solution_provider_id {
+            if SolutionProviders::<T>::get(sp_id).is_none() {
+                debug!(
+                    " ⚠️    Contract (id: {}): solution provider (id: {}) not exists",
+                    contract_id, sp_id
+                );
+            }
+        }
+    }
+
+    debug!(
+        "🏁  Smart Contract pallet {:?} checking Contracts storage map END",
+        PalletVersion::<T>::get()
+    );
+}
 
-        let contracts_to_bill_count: u64 = ContractsToBillAt::<T>::iter().count() as u64;
-        info!(
-            "🔎 CleanBillingLoop pre migration: Number of existing billing loop indexes {:?}",
-            contracts_to_bill_count
+fn check_node_contract<T: Config>(node_id: u32, contract_id: u64, deployment_hash: HexHash) {
+    if let Some(_) = pallet_tfgrid::Nodes::<T>::get(node_id) {
```

Comment:
The contained value of the Some variant is not used, so this check only cares if the key exists at all, which according to the docs can be checked as:

```suggestion
    if pallet_tfgrid::Nodes::<T>::contains_key(node_id) {
```

Reply:
indeed
which is also equivalent to:
~~~rust
if pallet_tfgrid::Nodes::<T>::get(node_id).is_some() {
~~~
correct?